### ISSUE TITLE
Crash during table fuzzing: Guarantee failed: [val.has()]

### ISSUE THREAD

**Issue by Tryneus** [rethinkdb/rethinkdb#4736]
Implemented a new table fuzzer to attempt to weed out any v2.1.x bugs we haven't found yet, got the following crash.  This test did not use any changefeeds, but did use multiple client threads.

```
2015-08-21T00:41:56.982640554 16.384082s error: Error in ../src/rdb_protocol/datum.cc at line 1896:
2015-08-21T00:41:56.983683027 16.385124s error: Guarantee failed: [val.has()]
2015-08-21T00:41:56.984502630 16.385944s error: Backtrace:
2015-08-21T00:42:02.139539142 21.540985s error: Fri Aug 21 00:41:56 2015

1: rethinkdb_backtrace(void**, int) at rethinkdb_backtrace.cc:100
2: backtrace_t::backtrace_t() at backtrace.cc:203
3: lazy_backtrace_formatter_t::lazy_backtrace_formatter_t() at backtrace.cc:283
4: format_backtrace(bool) at backtrace.cc:198
5: report_fatal_error(char const*, int, char const*, ...) at errors.cc:83
6: ql::datum_object_builder_t::add(datum_string_t const&, ql::datum_t) at datum.cc:1896
7: ql::datum_object_builder_t::add(char const*, ql::datum_t) at datum.cc:1904
8: make_replacement_pair(ql::datum_t, ql::datum_t) at table_common.cc:17
9: real_reql_cluster_interface_t::reconfigure_internal(counted_t<ql::db_t const> const&, uuid_u const&, table_generate_config_params_t const&, bool, signal_t*, ql::datum_t*) at real_reql_cluster_interface.cc:616
10: real_reql_cluster_interface_t::table_reconfigure(counted_t<ql::db_t const>, name_string_t const&, table_generate_config_params_t const&, bool, signal_t*, ql::datum_t*, admin_err_t*) at real_reql_cluster_interface.cc:640
11: artificial_reql_cluster_interface_t::table_reconfigure(counted_t<ql::db_t const>, name_string_t const&, table_generate_config_params_t const&, bool, signal_t*, ql::datum_t*, admin_err_t*) at artificial_reql_cluster_interface.cc:262
12: ql::reconfigure_term_t::eval_impl_on_table_or_db(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t, counted_t<ql::db_t const> const&, boost::optional<name_string_t> const&) const at db_table.cc:539
13: ql::table_or_db_meta_term_t::eval_impl(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t) const at db_table.cc:400
14: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at op.cc:225
15: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at term.cc:350
16: ql::query_cache_t::ref_t::run(ql::env_t*, Response*) at query_cache.cc:301
17: ql::query_cache_t::ref_t::fill_response(Response*) at query_cache.cc:247
18: ql::run(ql::query_id_t&&, ql::protob_t<Query>, Response*, ql::query_cache_t*, signal_t*) at term.cc:269
19: rdb_query_server_t::run_query(ql::query_id_t&&, ql::protob_t<Query> const&, Response*, ql::query_cache_t*, signal_t*) at query_server.cc:49
20: void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::query_server_t::connection_loop<json_protocol_t>::{lambda(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*)#1}::operator()(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*) const::{lambda()#1}::operator()() const at protob.cc:615
21: void save_exception<void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::query_server_t::connection_loop<json_protocol_t>::{lambda(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*)#1}::operator()(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*) const::{lambda()#1}>(std::__exception_ptr::exception_ptr*, std::string*, cond_t*, void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::query_server_t::connection_loop<json_protocol_t>::{lambda(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*)#1}::operator()(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*) const::{lambda()#1}&&) at protob.cc:555
22: void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::query_server_t::connection_loop<json_protocol_t>::{lambda(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*)#1}::operator()(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*) const at protob.cc:612
23: std::_Function_handler<void (std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*), void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::{lambda(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*)#1}>::_M_invoke(std::_Any_data const&, std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*) at functional:1779
24: std::function<void (std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*)>::operator()(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*) const at functional:2160
25: std_function_callback_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >::coro_pool_callback(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*) at coro_pool.hpp:30
26: coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >::worker_run(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, auto_drainer_t::lock_t) at coro_pool.hpp:68
27: std::_Mem_fn<void (coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >::*)(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, auto_drainer_t::lock_t)>::operator()(coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >*, std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, auto_drainer_t::lock_t) const at functional:550
28: void std::_Bind<std::_Mem_fn<void (coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >::*)(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, auto_drainer_t::lock_t)> (coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >*, std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, auto_drainer_t::lock_t)>::__call<void, , 0, 1, 2>(std::tuple<>&&, std::_Index_tuple<0, 1, 2>) at functional:1145
29: void std::_Bind<std::_Mem_fn<void (coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >::*)(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, auto_drainer_t::lock_t)> (coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >*, std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, auto_drainer_t::lock_t)>::operator()<, void>() at functional:1203
30: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >::*)(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, auto_drainer_t::lock_t)> (coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >*, std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, auto_drainer_t::lock_t)> >::run_action() at callable_action.hpp:31
31: callable_action_wrapper_t::run() at runtime_utils.cc:43
32: coro_t::run() at coroutines.cc:205
33: coro_t* coro_t::spawn_sometime<std::_Bind<std::_Mem_fn<void (coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >::*)(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, auto_drainer_t::lock_t)> (coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >*, std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, auto_drainer_t::lock_t)> >(std::_Bind<std::_Mem_fn<void (coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >::*)(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, auto_drainer_t::lock_t)> (coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >*, std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, auto_drainer_t::lock_t)>&&) at coroutines.hpp:58
34: coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >::on_source_availability_changed() at coro_pool.hpp:88
35: availability_control_t::set_available(bool) at passive_producer.hpp:65
36: limited_fifo_queue_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, std::list<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, std::allocator<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > > > >::push(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > const&) at limited_fifo.hpp:35
37: void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::query_server_t::connection_loop<json_protocol_t>::{lambda()#1}::operator()() const at protob.cc:649
38: void save_exception<void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::{lambda()#1}>(std::__exception_ptr::exception_ptr*, std::string*, cond_t*, void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::{lambda()#1}&&) at protob.cc:555
39: void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*) at protob.cc:644
40: query_server_t::handle_conn(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t) at protob.cc:489
41: std::_Mem_fn<void (query_server_t::*)(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t)>::operator()(query_server_t*, scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t) const at functional:550
42: void std::_Bind<std::_Mem_fn<void (query_server_t::*)(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t)> (query_server_t*, std::_Placeholder<1>, auto_drainer_t::lock_t)>::__call<void, scoped_ptr_t<linux_tcp_conn_descriptor_t>&, 0, 1, 2>(std::tuple<scoped_ptr_t<linux_tcp_conn_descriptor_t>&>&&, std::_Index_tuple<0, 1, 2>) at functional:1145
43: void std::_Bind<std::_Mem_fn<void (query_server_t::*)(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t)> (query_server_t*, std::_Placeholder<1>, auto_drainer_t::lock_t)>::operator()<scoped_ptr_t<linux_tcp_conn_descriptor_t>&, void>(scoped_ptr_t<linux_tcp_conn_descriptor_t>&) at functional:1203
44: std::_Function_handler<void (scoped_ptr_t<linux_tcp_conn_descriptor_t>&), std::_Bind<std::_Mem_fn<void (query_server_t::*)(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t)> (query_server_t*, std::_Placeholder<1>, auto_drainer_t::lock_t)> >::_M_invoke(std::_Any_data const&, scoped_ptr_t<linux_tcp_conn_descriptor_t>&) at functional:1779
45: std::function<void (scoped_ptr_t<linux_tcp_conn_descriptor_t>&)>::operator()(scoped_ptr_t<linux_tcp_conn_descriptor_t>&) const at functional:2160
46: linux_nonthrowing_tcp_listener_t::handle(int) at network.cc:924
47: std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(int)>::operator()(linux_nonthrowing_tcp_listener_t*, int) const at functional:550
48: void std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(int)> (linux_nonthrowing_tcp_listener_t*, int)>::__call<void, , 0, 1>(std::tuple<>&&, std::_Index_tuple<0, 1>) at functional:1145
49: void std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(int)> (linux_nonthrowing_tcp_listener_t*, int)>::operator()<, void>() at functional:1203
50: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(int)> (linux_nonthrowing_tcp_listener_t*, int)> >::run_action() at callable_action.hpp:31
51: callable_action_wrapper_t::run() at runtime_utils.cc:43
52: coro_t::run() at coroutines.cc:205
53: void coro_t::spawn_now_dangerously<std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(int)> (linux_nonthrowing_tcp_listener_t*, int)> >(std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(int)> (linux_nonthrowing_tcp_listener_t*, int)>&&) at coroutines.hpp:52
54: linux_nonthrowing_tcp_listener_t::accept_loop(auto_drainer_t::lock_t) at network.cc:891
55: std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(auto_drainer_t::lock_t)>::operator()(linux_nonthrowing_tcp_listener_t*, auto_drainer_t::lock_t) const at functional:550
56: void std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(auto_drainer_t::lock_t)> (linux_nonthrowing_tcp_listener_t*, auto_drainer_t::lock_t)>::__call<void, , 0, 1>(std::tuple<>&&, std::_Index_tuple<0, 1>) at functional:1145
57: void std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(auto_drainer_t::lock_t)> (linux_nonthrowing_tcp_listener_t*, auto_drainer_t::lock_t)>::operator()<, void>() at functional:1203
58: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(auto_drainer_t::lock_t)> (linux_nonthrowing_tcp_listener_t*, auto_drainer_t::lock_t)> >::run_action() at callable_action.hpp:31
59: callable_action_wrapper_t::run() at runtime_utils.cc:43
60: coro_t::run() at coroutines.cc:205
61: coro_t* coro_t::spawn_sometime<std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(auto_drainer_t::lock_t)> (linux_nonthrowing_tcp_listener_t*, auto_drainer_t::lock_t)> >(std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(auto_drainer_t::lock_t)> (linux_nonthrowing_tcp_listener_t*, auto_drainer_t::lock_t)>&&) at coroutines.hpp:58
62: linux_nonthrowing_tcp_listener_t::begin_listening() at network.cc:684
63: linux_tcp_listener_t::linux_tcp_listener_t(std::set<ip_address_t, std::less<ip_address_t>, std::allocator<ip_address_t> > const&, int, std::function<void (scoped_ptr_t<linux_tcp_conn_descriptor_t>&)> const&) at network.cc:955
64: query_server_t::query_server_t(rdb_context_t*, std::set<ip_address_t, std::less<ip_address_t>, std::allocator<ip_address_t> > const&, int, query_handler_t*, unsigned int) at protob.cc:369
65: rdb_query_server_t::rdb_query_server_t(std::set<ip_address_t, std::less<ip_address_t>, std::allocator<ip_address_t> > const&, int, rdb_context_t*) at query_server.cc:19
66: do_serve(io_backender_t*, bool, base_path_t const&, metadata_file_t*, serve_info_t const&, os_signal_cond_t*) at serve.cc:487
67: serve(io_backender_t*, base_path_t const&, metadata_file_t*, serve_info_t const&, os_signal_cond_t*) at serve.cc:635
68: run_rethinkdb_serve(base_path_t const&, serve_info_t*, file_direct_io_mode_t, int, boost::optional<boost::optional<unsigned long> > const&, uuid_u const*, server_config_versioned_t const*, cluster_semilattice_metadata_t const*, directory_lock_t*, bool*) at command_line.cc:805
69: void std::_Bind<void (*(base_path_t, serve_info_t*, file_direct_io_mode_t, int, boost::optional<boost::optional<unsigned long> >, uuid_u*, server_config_versioned_t*, cluster_semilattice_metadata_t*, directory_lock_t*, bool*))(base_path_t const&, serve_info_t*, file_direct_io_mode_t, int, boost::optional<boost::optional<unsigned long> > const&, uuid_u const*, server_config_versioned_t const*, cluster_semilattice_metadata_t const*, directory_lock_t*, bool*)>::__call<void, , 0, 1, 2, 3, 4, 5, 6, 7, 8, 9>(std::tuple<>&&, std::_Index_tuple<0, 1, 2, 3, 4, 5, 6, 7, 8, 9>) at functional:1145
70: void std::_Bind<void (*(base_path_t, serve_info_t*, file_direct_io_mode_t, int, boost::optional<boost::optional<unsigned long> >, uuid_u*, server_config_versioned_t*, cluster_semilattice_metadata_t*, directory_lock_t*, bool*))(base_path_t const&, serve_info_t*, file_direct_io_mode_t, int, boost::optional<boost::optional<unsigned long> > const&, uuid_u const*, server_config_versioned_t const*, cluster_semilattice_metadata_t const*, directory_lock_t*, bool*)>::operator()<, void>() at functional:1203
71: std::_Function_handler<void (), std::_Bind<void (*(base_path_t, serve_info_t*, file_direct_io_mode_t, int, boost::optional<boost::optional<unsigned long> >, uuid_u*, server_config_versioned_t*, cluster_semilattice_metadata_t*, directory_lock_t*, bool*))(base_path_t const&, serve_info_t*, file_direct_io_mode_t, int, boost::optional<boost::optional<unsigned long> > const&, uuid_u const*, server_config_versioned_t const*, cluster_semilattice_metadata_t const*, directory_lock_t*, bool*)> >::_M_invoke(std::_Any_data const&) at functional:1779
72: std::function<void ()>::operator()() const at functional:2160
73: starter_t::run_wrapper(std::function<void ()> const&) at runtime.cc:61
74: std::_Mem_fn<void (starter_t::*)(std::function<void ()> const&)>::operator()(starter_t*, std::function<void ()> const&) const at functional:550
75: void std::_Bind<std::_Mem_fn<void (starter_t::*)(std::function<void ()> const&)> (starter_t*, std::function<void ()>)>::__call<void, , 0, 1>(std::tuple<>&&, std::_Index_tuple<0, 1>) at functional:1145
76: void std::_Bind<std::_Mem_fn<void (starter_t::*)(std::function<void ()> const&)> (starter_t*, std::function<void ()>)>::operator()<, void>() at functional:1203
77: std::_Function_handler<void (), std::_Bind<std::_Mem_fn<void (starter_t::*)(std::function<void ()> const&)> (starter_t*, std::function<void ()>)> >::_M_invoke(std::_Any_data const&) at functional:1779
78: std::function<void ()>::operator()() const at functional:2160
79: callable_action_instance_t<std::function<void ()>&>::run_action() at callable_action.hpp:31
80: callable_action_wrapper_t::run() at runtime_utils.cc:43
81: coro_t::run() at coroutines.cc:205
82: coro_t* coro_t::spawn_sometime<std::function<void ()>&>(std::function<void ()>&) at coroutines.hpp:58
83: starter_t::on_thread_switch() at runtime.cc:57
84: linux_message_hub_t::on_event(int) at message_hub.cc:154
85: epoll_event_queue_t::run() at epoll.cc:115
86: linux_thread_pool_t::start_thread(void*) at thread_pool.cc:161
87: /lib/x86_64-linux-gnu/libpthread.so.0(+0x7e9a) [0x7f5819814e9a] at 0x7f5819814e9a (/lib/x86_64-linux-gnu/libpthread.so.0)
88: clone+0x6d at 0x7f58195418bd (/lib/x86_64-linux-gnu/libc.so.6)
2015-08-21T00:42:02.143574006 21.545015s error: Exiting.
```

**Comment by Tryneus** [comment]
Just got a similar crash in `make_replacement_pair` from `table_rebalance` instead of `table_reconfigure`, in case it matters:

```
2015-08-21T00:52:31.749917561 78.132890s error: Error in ../src/rdb_protocol/datum.cc at line 1896:
2015-08-21T00:52:31.751461679 78.134435s error: Guarantee failed: [val.has()]
2015-08-21T00:52:31.752000482 78.134973s error: Backtrace:
2015-08-21T00:52:37.448497328 83.831474s error: Fri Aug 21 00:52:31 2015

1: rethinkdb_backtrace(void**, int) at rethinkdb_backtrace.cc:100
2: backtrace_t::backtrace_t() at backtrace.cc:203
3: lazy_backtrace_formatter_t::lazy_backtrace_formatter_t() at backtrace.cc:283
4: format_backtrace(bool) at backtrace.cc:198
5: report_fatal_error(char const*, int, char const*, ...) at errors.cc:83
6: ql::datum_object_builder_t::add(datum_string_t const&, ql::datum_t) at datum.cc:1896
7: ql::datum_object_builder_t::add(char const*, ql::datum_t) at datum.cc:1904
8: make_replacement_pair(ql::datum_t, ql::datum_t) at table_common.cc:17
9: real_reql_cluster_interface_t::rebalance_internal(uuid_u const&, signal_t*, ql::datum_t*) at real_reql_cluster_interface.cc:855
10: real_reql_cluster_interface_t::table_rebalance(counted_t<ql::db_t const>, name_string_t const&, signal_t*, ql::datum_t*, admin_err_t*) at real_reql_cluster_interface.cc:872
11: artificial_reql_cluster_interface_t::table_rebalance(counted_t<ql::db_t const>, name_string_t const&, signal_t*, ql::datum_t*, admin_err_t*) at artificial_reql_cluster_interface.cc:316
12: ql::rebalance_term_t::eval_impl_on_table_or_db(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t, counted_t<ql::db_t const> const&, boost::optional<name_string_t> const&) const at db_table.cc:611
13: ql::table_or_db_meta_term_t::eval_impl(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t) const at db_table.cc:400
14: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at op.cc:225
15: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at term.cc:350
16: ql::query_cache_t::ref_t::run(ql::env_t*, Response*) at query_cache.cc:301
17: ql::query_cache_t::ref_t::fill_response(Response*) at query_cache.cc:247
18: ql::run(ql::query_id_t&&, ql::protob_t<Query>, Response*, ql::query_cache_t*, signal_t*) at term.cc:269
```

**Comment by VeXocide** [comment]
I'll take this one

**Comment by VeXocide** [comment]
`common_table_artificial_table_backend_t::read_row` will return a default-constructed `ql::datum_t` and `true` when it hits an `no_such_table_exc_t` exception in formatting the row, which is likely what's going on here. There's a few ways of fixing this, 'll need to check which makes the most sense.

**Comment by VeXocide** [comment]
For clarity, this is probably the behaviour we want in the artificial tables, but not in the `real_reql_cluster_interface_t`.

**Comment by VeXocide** [comment]
While a `!row.has()` gets rewritten into a `ql::datum_t::null()` in `artificial_table_t::read_row` the function `checked_read_row_from_backend` in there depends on the difference.

**Comment by VeXocide** [comment]
Merged into branches `next` and `v2.1.x` via commits f3c5d64a075ab004a9944b0223b5e2ec4e342b22 and b5234a0f1625f44d14feecb08f1e2c377f7ebddf respectively.
